src/platforms/generic_unix/socket_driver.c

```diff
@@ -0,0 +1,253 @@
+
+#include "atom.h"
+#include "context.h"
+#include "ccontext.h"
+#include "globalcontext.h"
+#include "interop.h"
+#include "utils.h"
+#include "term.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/udp.h>
+#include <errno.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "trace.h"
+#include "sys.h"
+
+#define BUFSIZE 1024
+
+static const char *const tag_proto_a = "\x5" "proto";
+static const char *const proto_udp_a = "\x3" "udp";
+static const char *const proto_tcp_a = "\x3" "tcp";
+
+typedef struct SocketDriverData
+{
+    int sockfd;
+} SocketDriverData;
+
+
+void *socket_driver_create_data()
+{
+    struct SocketDriverData *data = calloc(1, sizeof(struct SocketDriverData));
+    return (void *) data;
+}
+
+
+void socket_driver_delete_data(void *data)
+{
+    free(data);
+}
+
+
+term socket_driver_do_init(CContext *cc, term params)
+{
+    Context *ctx = cc->ctx;
+    SocketDriverData *socket_data = (SocketDriverData *) ctx->platform_data;
+
+    if (!term_is_list(params)) {
+        return port_create_error_tuple(cc, "badarg: params is not a list");
+    }
+    term proto = interop_proplist_get_value(params, context_make_atom(ctx, tag_proto_a));
+
+    if (term_is_nil(proto)) {
+        return port_create_error_tuple(cc, "badarg: no proto in params");
+    }
+
+    if (proto == context_make_atom(ctx, proto_udp_a)) {
+        int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+        if (sockfd == -1) {
+            const char *error_string = strerror(errno);
+            return port_create_error_tuple(cc, error_string);
+        }
+        socket_data->sockfd = sockfd;
+    } else if (proto == context_make_atom(ctx, proto_tcp_a)) {
+        socket_data->sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    } else {
+        return port_create_error_tuple(cc, "badarg: unsupported protocol");
+    }
+    if (fcntl(socket_data->sockfd, F_SETFL, O_NONBLOCK) == -1){
+        const char* error_string = strerror(errno);
+        return port_create_error_tuple(cc, error_string);
+    }
+
+    return context_make_atom(ctx, port_ok_a);
+}
+
+
+term socket_driver_do_bind(CContext *cc, term address, term port)
+{
+    Context *ctx = cc->ctx;
+    SocketDriverData *socket_data = (SocketDriverData *) ctx->platform_data;
+
+    struct sockaddr_in serveraddr;
+
+    UNUSED(address);
+    memset(&serveraddr, 0, sizeof(serveraddr));
+    serveraddr.sin_family = AF_INET;
+    serveraddr.sin_addr.s_addr = htonl(INADDR_ANY); // TODO
+    serveraddr.sin_port = htons(term_to_int32(port));
+
+    socklen_t address_len = sizeof(serveraddr);
+    if (bind(socket_data->sockfd, (struct sockaddr *) &serveraddr, address_len) == -1) {
+        const char* error_string = strerror(errno);
+        return port_create_error_tuple(cc, error_string);
+    } else {
+        if (getsockname(socket_data->sockfd, (struct sockaddr *) &serveraddr, &address_len) == -1) {
+            const char* error_string = strerror(errno);
```

Comment:
it should be: `const char *error_string`. am I reviewing an outdated version?

Reply:
fixed